ssh exporter (#2138)
grafana/alloy: docs/sources/reference/components/prometheus/prometheus.exporter.ssh.md

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.ssh/
description: Learn about prometheus.exporter.ssh
labels:
  stage: experimental
  products:
    - oss
title: prometheus.exporter.ssh
---

# `prometheus.exporter.ssh`

{{< docs/shared lookup="stability/experimental.md" source="alloy" version="<ALLOY_VERSION>" >}}

The `prometheus.exporter.ssh` component embeds an SSH exporter for collecting metrics from remote servers over SSH and exporting them as Prometheus metrics.

## Usage

```alloy
prometheus.exporter.ssh "example" {
  targets {
    address         = "192.168.1.10"
    username        = "admin"
    key_file     = "path/to/key.pem"
    custom_metrics {
      name    = "load_average"
      command = "cat /proc/loadavg | awk '{print $1}'"
      type    = "gauge"
      help    = "Load average over 1 minute"
      labels  = { host = "192.168.1.10" }  // attach host as a label
    }
  }
}

> 
> 

Reviewer: ```suggestion

# Arguments

The `prometheus.exporter.ssh` component doesn't support any arguments. You can configure this component with blocks.

```

## Blocks

You can use the following blocks with `prometheus.exporter.ssh`:

| Block                                          | Description                                     | Required |
| ---------------------------------------------- | ----------------------------------------------- | -------- |
| [`targets`][targets]                           | Configures SSH targets to collect metrics from. | yes      |
| `targets` > [`custom_metrics`][custom_metrics] | Defines metrics to collect from a server.       | yes      |

The `>` symbol indicates deeper levels of nesting.
For example, `targets` > `custom_metrics` refers to a `custom_metrics` block defined inside a `target` block.

[targets]: #targets
[custom_metrics]: #custom_metrics

### `targets`

> Configures SSH targets to collect metrics from.

Reviewer: ```suggestion
<span class="badge docs-labels__stage docs-labels__item">Required</span>

Configures SSH targets to collect metrics from.
```

The following arguments are supported:

| Name              | Type       | Description                                        | Default | Required |
| ----------------- | ---------- | -------------------------------------------------- | ------- | -------- |
| `address`         | `string`   | The IP or hostname of the target server.           |         | yes      |
| `username`        | `string`   | SSH username.                                      |         | yes      |
| `command_timeout` | `duration` | Timeout for each SSH command.                      | `"30s"` | no       |
| `key_file`        | `string`   | Private key file path for key-based authorization. |         | no       |
| `password`        | `secret`   | Password for SSH login.                            |         | no       |
| `port`            | `int`      | SSH port number.                                   | `22`    | no       |

Either `password` or `key_file` must be set. If both are provided, `key_file` is used.

### `custom_metrics`

> Defines metrics to collect from a server.

Reviewer: ```suggestion
Defines metrics to collect from a server.

<span class="badge docs-labels__stage docs-labels__item">Required</span>

```

The following arguments are supported:

You can attach arbitrary labels to your metrics using the `labels` argument in `custom_metrics` (for example, `labels = { address = each.value }`).

| Name          | Type                  | Description                                              | Default | Required |
| ------------- | --------------------- | -------------------------------------------------------- | ------- | -------- |
| `command`     | `string`              | Command to run remotely to get the metric value.         |         | yes      |
| `name`        | `string`              | Name of the exported metric.                             |         | yes      |
| `type`        | `string`              | Metric type: `gauge` or `counter`.                       |         | yes      |
| `help`        | `string`              | Help text for the metric.                                |         | no       |
| `labels`      | `map(string, string)` | Additional labels to attach to the metric.               | `{}`    | no       |
| `parse_regex` | `string`              | Regular expression to extract value from command output. |         | no       |

> ## Secure Known Hosts Setup

Reviewer: ```suggestion
## Secure known hosts setup
```

> 
> ### How It Works

Reviewer: ```suggestion

{{< param "PRODUCT_NAME" >}} performs the following steps to set up known hosts:
```

1. **First Run**: If `~/.ssh/known_hosts` is missing, a new one is created using `ssh-keyscan`.

> 2. **Validation**: Host keys are validated on every connection attempt.
> 3. **Changes**: Key mismatches raise an error requiring manual review.
> 4. **New Targets**: Automatically scanned and added, but mismatches block the connection.

Reviewer: ```suggestion
1. **Validation**: Host keys are validated on every connection attempt.
1. **Changes**: Key mismatches raise an error requiring manual review.
1. **New Targets**: Automatically scanned and added, but mismatches block the connection.
```

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.ssh/
aliases:
  - ../prometheus.exporter.ssh/ # /docs/alloy/latest/reference/components/prometheus.exporter.ssh/
description: Learn about prometheus.exporter.ssh
title: prometheus.exporter.ssh
---

# prometheus.exporter.ssh

The `prometheus.exporter.ssh` component embeds an SSH exporter for collecting metrics from remote servers over SSH and exporting them as Prometheus metrics.

> 

Reviewer 2: Please add the experimental snippet to document the fact that the component is experimental

## Usage

```
prometheus.exporter.ssh "LABEL" {
  // Configuration options
}
```

## Arguments

The following arguments can be used to configure the exporter's behavior.

> All arguments are optional unless specified. Omitted fields take their default values.

Reviewer: Observation: In reviewing this I've noticed that the wording of some sections in the Prometheus topics are not consistent with the rest of the components. For example, Arguments in other sections simply state: `The following arguments are supported:`

I'll create an Issue to follow up on this.  No action or suggestions for this PR.

| Name              | Type     | Description                                        | Default | Required |
| ----------------- | -------- | -------------------------------------------------- | ------- | -------- |

> | `verbose_logging` | `bool`   | Enable verbose logging for debugging purposes.     | `false` | no       |

Reviewer 3: Do we really need such a argument? Normally users would configure the logging level through the [logging block](https://grafana.com/docs/alloy/latest/reference/config-blocks/logging/).

| `targets`         | `block`  | One or more target configurations for SSH metrics. |         | yes      |

## Blocks

The following blocks are supported inside the definition of `prometheus.exporter.ssh`:

> | Block          | Description                                                 | Required |

Reviewer: Should we add/include the `Hierarchy` column in this table?

### targets block

The `targets` block defines the remote servers to connect to and the metrics to collect. It supports the following arguments:

| Name              | Type                  | Description                                                            | Default | Required |
| ----------------- | --------------------- | ---------------------------------------------------------------------- | ------- | -------- |
| `address`         | `string`              | The IP address or hostname of the target server.                       |         | yes      |
| `port`            | `int`                 | SSH port number.                                                       | `22`    | no       |
| `username`        | `string`              | SSH username for authentication.                                       |         | yes      |
| `password`        | `secret`              | Password for password-based SSH authentication.                        |         | Required if `key_file` is not provided |
| `key_file`        | `string`              | Path to the private key file for key-based SSH authentication.         |         | Required if `password` is not provided |

> | `command_timeout` | `int`                 | Timeout in seconds for each command execution over SSH.                | `30`    | no       |

Reviewer 4: Should this be a duration? We rarely use dedicated seconds.

Reviewer 4: What is the expected behavior if command timeout > scrape timeout?

| `name`         | `string`              | The name of the metric.                                                      |         | yes      |
| `command`      | `string`              | The command to execute over SSH to collect the metric.                       |         | yes      |
| `type`         | `string`              | The type of the metric (`gauge` or `counter`).                               |         | yes      |
| `help`         | `string`              | Help text for the metric.                                                    |         | no       |
| `labels`       | `map(string, string)` | Key-value pairs of labels to associate with the metric.                      | `{}`    | no       |
| `parse_regex`  | `string`              | Regular expression to parse the command output and extract the metric value. |         | no       |

#### Metric Types

- `gauge`: Represents a numerical value that can go up or down.
- `counter`: Represents a cumulative value that only increases.

#### parse_regex

If the command output is not a simple numeric value, use `parse_regex` to extract the numeric value from the output.

---

## Secure Known Hosts Setup

> ### How It Works

Reviewer 3: That's true, although there are some components such as [this](https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.remote_write/#technical-details) which have similar sections.

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.ssh/
description: Learn about prometheus.exporter.ssh
labels:
  stage: experimental
  products:
    - oss
title: prometheus.exporter.ssh
---

# `prometheus.exporter.ssh`

{{< docs/shared lookup="stability/experimental.md" source="alloy" version="<ALLOY_VERSION>" >}}

The `prometheus.exporter.ssh` component embeds an SSH exporter for collecting metrics from remote servers over SSH and exporting them as Prometheus metrics.

## Usage

```alloy
prometheus.exporter.ssh "example" {
  targets {
    address         = "192.168.1.10"
    username        = "admin"
    key_file     = "path/to/key.pem"
    custom_metrics {
      name    = "load_average"
      command = "cat /proc/loadavg | awk '{print $1}'"
      type    = "gauge"
      help    = "Load average over 1 minute"
      labels  = { host = "192.168.1.10" }  // attach host as a label
    }
  }

> }
> 

Reviewer: ````suggestion
}
```

````

---
canonical: https://grafana.com/docs/alloy/latest/reference/components/prometheus/prometheus.exporter.ssh/
description: Learn about prometheus.exporter.ssh
title: prometheus.exporter.ssh
---

> # prometheus.exporter.ssh

Reviewer: ```suggestion
# `prometheus.exporter.ssh`
```

{{< docs/shared lookup="stability/experimental.md" source="alloy" version="<ALLOY_VERSION>" >}}

The `prometheus.exporter.ssh` component embeds an SSH exporter for collecting metrics from remote servers over SSH and exporting them as Prometheus metrics.

> ## Usage

Reviewer 3: Yes, the usage section should be reworked to only show the required arguments and blocks. It should not show optional ones.

      help    = "Load average over 1 minute" // optional help text
    }
  }
}
```

## Example: Curated Targets via Discovery

```alloy
locals {
  ssh_keys = {
    for path in filesystem.glob("${path.module}/keys/*.pem") : basename(path, ".pem") => path
  }
}

prometheus.exporter.ssh "curated" {
  # Iterate only over discovered hosts with a matching key
  for_each = data.discovery.hosts.byLabel("app=backend")

  targets {
    address    = each.value             // host or IP from discovery
    username   = "monitor"             // required: SSH user
    key_file   = ssh_keys[each.value]    // only hosts with key files

    custom_metrics {
      name    = "uptime"               // required: metric name
      command = "cat /proc/uptime | awk '{print $1}'"  // required: command
      type    = "gauge"                // required: gauge or counter
    }
  }
}
```

## Arguments

| Name              | Type   | Description                                    | Default | Required |
|-------------------|--------|------------------------------------------------|---------|----------|
| `verbose_logging` | `bool` | Enable verbose logging for debugging purposes. | `false` | no       |

> ## Blocks
> 

Reviewer: ```suggestion
## Blocks

You can use the following blocks with `prometheus.exporter.ssh`:

| Block                                          | Description                                     | Required |
| ---------------------------------------------- | ----------------------------------------------- | -------- |
| [`targets`][targets]                           | Configures SSH targets to collect metrics from. | yes      |
| `targets` > [`custom_metrics`][custom_metrics] | Defines metrics to collect from a server.       | yes      |

The `>` symbol indicates deeper levels of nesting.
For example, `targets` > `custom_metrics` refers to a `custom_metrics` block defined inside a `target` block.

[targets]: #targets
[custom_metrics]: #custom_metrics

```

> ### `targets` block

Reviewer: ```suggestion
### `targets`
```

|-------------------|-----------------------|------------------------------------------------------------------------|---------|----------|
| `address`         | `string`              | The IP or hostname of the target server.                               |         | yes      |
| `port`            | `int`                 | SSH port number.                                                       | `22`    | no       |
| `username`        | `string`              | SSH username.                                                          |         | yes      |
| `password`        | `secret`              | Password for SSH login.                                                |         | no       |
| `key_file`        | `string`              | Private key file path for key-based auth.                              |         | no       |
| `command_timeout` | `duration`            | Timeout for each SSH command.                                          | `30s`   | no       |
| `custom_metrics`  | `block`               | One or more metrics to collect via SSH.                                |         | yes      |

> > Either `password` or `key_file` must be set. If both are provided, `key_file` is used.

Reviewer: ```suggestion
Either `password` or `key_file` must be set. If both are provided, `key_file` is used.
```

> ### `custom_metrics` block

Reviewer: ```suggestion
### `custom_metrics`
```

> Defines metrics to collect from a server.

Reviewer: ```suggestion
Defines metrics to collect from a server.

The following arguments are supported:
```

> | Name           | Type                  | Description                                                                  | Default | Required |
> |----------------|-----------------------|------------------------------------------------------------------------------|---------|----------|
> | `name`         | `string`              | Name of the exported metric.                                                |         | yes      |
> | `command`      | `string`              | Command to run remotely to get the metric value.                            |         | yes      |
> | `type`         | `string`              | Metric type: `gauge` or `counter`.                                          |         | yes      |
> | `help`         | `string`              | Help text for the metric.                                                   |         | no       |
> | `labels`       | `map(string, string)` | Additional labels to attach to the metric.                                  | `{}`    | no       |
> | `parse_regex`  | `string`              | Regex to extract value from command output.                                 |         | no       |

Reviewer: ```suggestion
| Name          | Type                  | Description                                              | Default | Required |
| ------------- | --------------------- | -------------------------------------------------------- | ------- | -------- |
| `command`     | `string`              | Command to run remotely to get the metric value.         |         | yes      |
| `name`        | `string`              | Name of the exported metric.                             |         | yes      |
| `type`        | `string`              | Metric type: `gauge` or `counter`.                       |         | yes      |
| `help`        | `string`              | Help text for the metric.                                |         | no       |
| `labels`      | `map(string, string)` | Additional labels to attach to the metric.               | `{}`    | no       |
| `parse_regex` | `string`              | Regular expression to extract value from command output. |         | no       |
```

> ## Secure Known Hosts Setup

Reviewer: Is this section a detailed example? An application scenario or usecase?

### How It Works

> 1. **First Run**: If `~/.ssh/known_hosts` is missing, a new one is created using `ssh-keyscan`.

Reviewer: I suspect this section will be reworked a little - it's not typical to use steps for an explanation of how things work. I want to review this after the rest of the doc suggestions are sorted out :-)

2. **Validation**: Host keys are validated on every connection attempt.
3. **Changes**: Key mismatches raise an error requiring manual review.
4. **New Targets**: Automatically scanned and added, but mismatches block the connection.

> ### Manual Resolution

Reviewer: ```suggestion
### Manual resolution
```

Use `ssh-keyscan` or another secure method to update `known_hosts` when a host key legitimately changes.

> ## Security Considerations

Reviewer: ```suggestion
## Security considerations
```

- Only valid IPs/hostnames accepted.
- Backticks and semicolons are disallowed in commands to prevent injection.
- SSH key files must be `0600`.
- Known-hosts entries are bootstrapped and never overwritten.
- On Windows, host-key checks are skipped.

## Exported fields

{{< docs/shared lookup="reference/components/exporter-component-exports.md" source="alloy" version="<ALLOY_VERSION>" >}}

## Component health

> `prometheus.exporter.ssh` is only reported as unhealthy if given an invalid configuration. In those cases, exported fields retain their last healthy values.

Reviewer: ```suggestion
`prometheus.exporter.ssh` is only reported as unhealthy if given an invalid configuration.
In those cases, exported fields retain their last healthy values.
```

## Debug metrics

`prometheus.exporter.ssh` doesn't expose any component-specific debug metrics.

## Example: Prometheus Scrape

```alloy
prometheus.exporter.ssh "example" {
  verbose_logging = true

  targets {
    address         = "192.168.1.10"
    port            = 22
    username        = "admin"
    password        = "password"
    command_timeout = "10s"

    custom_metrics {
      name    = "load_average"
      command = "cat /proc/loadavg | awk '{print $1}'"
      type    = "gauge"
      help    = "Load average over 1 minute"
    }
  }

  targets {
    address         = "192.168.1.11"
    port            = 22
    username        = "monitor"
    key_file        = "/path/to/private.key"
    command_timeout = "15s"

    custom_metrics {
      name        = "disk_usage"
      command     = "df / | tail -1 | awk '{print $5}'"
      type        = "gauge"
      help        = "Disk usage percentage"
      parse_regex = "(\d+)%"
    }
  }
}

prometheus.scrape "demo" {
  targets    = prometheus.exporter.ssh.example.targets
  forward_to = [prometheus.remote_write.demo.receiver]
}

prometheus.remote_write "demo" {
  endpoint {

>     url = PROMETHEUS_REMOTE_WRITE_URL
> 
>     basic_auth {
>       username = USERNAME
>       password = PASSWORD

Reviewer: ```suggestion
    url = <PROMETHEUS_REMOTE_WRITE_URL>

    basic_auth {
      username = <USERNAME>
      password = <PASSWORD>
```

    }
  }
}
```

Replace the following:

> - `PROMETHEUS_REMOTE_WRITE_URL`: Remote write-compatible server URL.
> - `USERNAME`: Auth username.
> - `PASSWORD`: Auth password.

Reviewer: ```suggestion
* _`<PROMETHEUS_REMOTE_WRITE_URL>`_: Remote write-compatible server URL.
* _`<USERNAME>`_: Auth username.
* _`<PASSWORD>`_: Auth password.
```

What is `Auth` in this context? We shouldn't use shortcuts like `auth` when describing or defining something. It's ambiguous and leaves it up to the reader to try to guess.

For example is USERNAME best described as:
```
The username to use for authentication to the `remote_write` API.
```
